docs/docs/user_guide/env_variables.md
# Environment Variables

The docker app requieres some **environment variables** to specify how to wire things up on launch. These variables are located 
in environment `.env` files and every component of the app requires a specific environment. In this section we will explain which `.env` files
and variables we have and good defaults for a quick test launch.

We use the following environment files:

* `.env` : Provides config for the web server, which implements the main app.
* `.env.celery` : Configs for celery, the async task backend.
* `.env.db` : Configs for the postgres database backend.

## Default files

I this section we will explain about each file and its corresponding environment variables, and we will provide some useful default values 
you can use as well. 

!!! Note 
    You might need a Django secret key, you can generate one using [this site](https://djecrety.ir). 

### .env

```
DEBUG=1        # If this is a debug deployment, 1 for true, 0 for false
SECRET_KEY=your_django_secret_key            # Change with your own secret key
DJANGO_ALLOWED_HOSTS=localhost 127.0.0.1 [::1] nginx        
CSRF_TRUSTED_ORIGINS=http://127.0.0.1:1337 https://localhost:1337 http://localhost:1337
SQL_ENGINE=django.db.backends.postgresql
SQL_DATABASE=early_warnings            # Change for your database name (the one you refer to when you connect to a DB)
SQL_USER=your_user_name                     # Change for your database username
SQL_PASSWORD=your_user_pass                 # Change for your database user password
SQL_HOST=postgres                           
SQL_PORT=5432
DATABASE=postgres
CELERY_BROKER_URL=redis://redis:6379
CELERY_RESULT_BACKEND=redis://redis:6379
```

### .env.celery

```
DEBUG=1                                     # If this is a debug deployment, 1 for true, 0 for false
SECRET_KEY=your_django_secret_key           # Change with your own secret key
DJANGO_ALLOWED_HOSTS=localhost 127.0.0.1 [::1] redis

CSRF_TRUSTED_ORIGINS=http://127.0.0.1:1337 https://localhost:1337 http://localhost:1337
SQL_ENGINE=django.db.backends.postgresql
SQL_DATABASE=early_warnings            # Change for your database name (the one you refer to when you connect to a DB)
SQL_USER=your_user_name                     # Change for your database username
SQL_PASSWORD=your_user_pass                 # Change for your database user password
SQL_HOST=postgres
SQL_PORT=5432
DATABASE=postgres
CELERY_BROKER_URL=redis://redis:6379
CELERY_RESULT_BACKEND=redis://redis:6379
```

### .env.db
POSTGRES_USER=your_user_name                # Change, Same as SQL_USER above
POSTGRES_PASSWORD=your_user_pass            # Change, Same as SQL_PASSWORD above
POSTGRES_DB=early_warnings             # Change, Same as SQL_DATABASE above



docker_project/early_warnings/early_warnings/early_warnings/settings.py
"""
Django settings for early_warnings project.

Generated by 'django-admin startproject' using Django 4.0.5.

For more information on this file, see
https://docs.djangoproject.com/en/4.0/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/4.0/ref/settings/
"""

from pathlib import Path
import os
# -- Init env variables handler -------------
import environ

env = environ.Env()
env.read_env()
# -------------------------------------------

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/4.0/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = env("SECRET_KEY")

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = int(env("DEBUG", default=1)) # type: ignore

ALLOWED_HOSTS = env("DJANGO_ALLOWED_HOSTS").split(" ")  # type: ignore
print(ALLOWED_HOSTS)
CSRF_TRUSTED_ORIGINS = ['http://localhost:1337'] 

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    'django_celery_beat',
    'blocking_early_warnings'
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'early_warnings.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates')],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'early_warnings.wsgi.application'


# Database
# https://docs.djangoproject.com/en/4.0/ref/settings/#databases

DATABASES = {
    "default": {
        "ENGINE": env("SQL_ENGINE", default="django.db.backends.sqlite3"),  # type: ignore
        "NAME": env("SQL_DATABASE", default=str(BASE_DIR / "db.sqlite3")),  # type: ignore
        "USER": env("SQL_USER", default="user"),  # type: ignore
        "PASSWORD": env("SQL_PASSWORD", default="password"),  # type: ignore
        "HOST": env("SQL_HOST", default="localhost"),  # type: ignore
        "PORT": env("SQL_PORT", default="5432"),  # type: ignore
    }
}


# Password validation
# https://docs.djangoproject.com/en/4.0/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/4.0/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/4.0/howto/static-files/

STATIC_URL = 'static/'
STATIC_ROOT = Path("/home/vsf/web/staticfiles")

# Default primary key field type
# https://docs.djangoproject.com/en/4.0/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'

# -- < Celery Settings > ---------------------------------

import early_warnings.tasks as tasks
from celery.schedules import crontab

CELERY_BROKER_URL = env("CELERY_BROKER_URL", default="redis://redis:6379") # type: ignore 
CELERY_RESULT_BACKEND = env("CELERY_RESULT_BACKEND", default="redis://redis:6379") # type: ignore

CELERY_BEAT_SCHEDULER = 'django_celery_beat.schedulers:DatabaseScheduler'

# --------------------------------------------------------

# -- < Login Settings > ----------------------------------

LOGIN_REDIRECT_URL = "/"
EARLY_WARNINGS_LOGIN_REQUIRED = True